akka-contrib/src/main/scala/akka/contrib/pattern/TransparentExponentialBackoffSupervisor.scala

```diff
@@ -0,0 +1,206 @@
+    Props(
+      classOf[TransparentExponentialBackoffSupervisor],
+      childProps,
+      None,
+      minBackoff,
+      maxBackoff,
+      randomFactor)
+  }
+}
+
+/**
+ * A supervising actor that restarts a child actor with an exponential back off.
+ *
+ * This explicit supervisor behaves similarly to the normal implicit supervision where
+ * if an actor throws an exception, the decider on the supervisor will decide when to
+ * `Stop`, `Restart`, `Escalate`, `Resume` the child actor.
+ *
+ * When the `Restart` directive is specified, the supervisor will delay the restart
+ * using an exponential back off strategy (bounded by minBackoff and maxBackoff).
+ *
+ * This supervisor is intended to be transparent to both the child actor and external actors.
+ * Where external actors can send messages to the supervisor as if it was the child and the
+ * messages will be forwarded. And when the child is `Terminated`, the supervisor is also
+ * `Terminated`.
+ * Transparent to the child means that the child does not have to be aware that it is being
+ * supervised specifically by the [[TransparentExponentialBackoffSupervisor]]. Just like it does
+ * not need to know when it is being supervised by the usual implicit supervisors.
+ * The only caveat is that the `ActorRef` of the child is not stable, so any user storing the
+ * `sender()` `ActorRef` from the child response may eventually not be able to communicate with
+ * the stored `ActorRef`. In general all messages to the child should be directed through the
+ * [[TransparentExponentialBackoffSupervisor]].
+ *
+ * An example of where this supervisor might be used is when you may have an actor that is
+ * responsible for continuously polling on a server for some resource that sometimes may be down.
+ * Instead of hammering the server continuously when the resource is unavailable, the actor will
+ * be restarted with an exponentially increasing back off until the resource is available again.
+ */
+class TransparentExponentialBackoffSupervisor(
+  props: Props,
+  decider: Option[Decider],
+  minBackoff: FiniteDuration,
+  maxBackoff: FiniteDuration,
+  randomFactor: Double)
+  extends Actor
+  with Stash
+  with ActorLogging {
+
+  import TransparentExponentialBackoffSupervisor._
+  import context._
+
+  override val supervisorStrategy = OneForOneStrategy(maxNrOfRetries = -1) {
+    case ex ⇒
+      val defaultDirective: Directive =
+        super.supervisorStrategy.decider.applyOrElse(ex, (_: Any) ⇒ Escalate)
+      val maybeDirective: Option[Directive] = decider
+        .map(_.applyOrElse(ex, (_: Any) ⇒ defaultDirective))
+
+      // Get the directive from the specified decider or fallback to
+      // the default decider.
+      // Whatever the final Directive is, we will translate all Restarts
+      // to our own Restarts, which involves stopping the child.
+      maybeDirective.getOrElse(defaultDirective) match {
+        case Restart ⇒
+          self ! RestartChild
+          Stop
```

Comment:
Yea, I was afraid that might be the case.
But thankfully, I actually had an alternate implementation before that did a `Resume` here instead of a `Stop` (still sending a `RestartChild` message) and then in the `RestartChild` handler, it would terminate the child before the become(waitingToRestart(childRef, numRestarts)).
I'll switch it to this implementation later tonight, but I was wondering if you could elaborate a little more on why the `Stop` can overtake the `RestartChild` message in the presence of other messages (just for my own edification).